### PR TITLE
Improve accessibility for issue comments

### PR DESCRIPTION
**Pull request by Menelion**
### Preamble

Gitea is an extremely great and smart solution perfectly suitable for smaller systems and self-hosted Git-powered setups. However, there is a group of people who have indredible difficulties in using Gitea, rendering it useless in many cases. Those people are blind developers using [screen readers](https://en.wikipedia.org/wiki/Screen_reader). Unfortunately, the frontend framework is super convoluted, and Go templates don’t allow accessibility improvements in a straightforward way. As a blind developer myself, I'm trying to start fixing smaller accessibility quirks with intention to go deeper and eventually, alone or not, make Gitea at least mostly accessible for screen reader users.

### What This Fix Does

My blind fellows and me navigate webpages not very similarly to how a sighted person does it. For instance, we love semantic HTML markup like headings, lists, tables etc. because our screen readers allow us to jump by those landmarks with a single keypress.  
Currently in Gitea issue comments are not marked up with headings. I'm trying to fix this by adding an appropriate [ARIA](https://www.w3.org/WAI/standards-guidelines/aria/) role for comment header and also by enclosing the comment itself in a semantical article element.

### CODE REVIEW

File: templates/repo/issue/view_content.tmpl
Comment (delvh) [inline templates/repo/issue/view_content.tmpl]: `complementary` for reactions?
This doesn't seem right, if I read the MDN docs correctly…

File: templates/repo/issue/view_content.tmpl
Comment (delvh) [inline templates/repo/issue/view_content.tmpl]: Hmm, is that really a heading?
I mean, it doesn't have any title-like attributes, and doesn't behave exactly like a title…

Comment (tallship) [reply templates/repo/issue/view_content.tmpl]: Isn't the objective just to use Aria article like accessibility for readers?

Comment (fsologureng) [reply templates/repo/issue/view_content.tmpl]: > Hmm, is that really a heading? I mean, it doesn't have any title-like attributes, and doesn't behave exactly like a title…

The whole container functions like a comment head, but change the container to a H3 it has a deep impact in visual presentation. That's why this proposition does an aria fix.

File: templates/repo/issue/view_content.tmpl
Comment (zeripath) [reply templates/repo/issue/view_content.tmpl]: probably should be `comment` AFAICS

Comment (fsologureng) [reply templates/repo/issue/view_content.tmpl]: I agree that [is not good for this](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/complementary_role#accessibility_concerns). For landmark roles «[it is not as good practice to use non-semantic elements, such as `<div>`, adding semantics with landmark roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/landmark_role#best_practices)».
Instead, I would suggest a [menu role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role).

**Comment by fsologureng** [comment]
@zeripath I don't know if you've read that @Menelion can't stop using a screen reader?

**Review by zeripath — APPROVED** [review]
If users using a screenreader say that this works then...

File: templates/repo/issue/view_content.tmpl
Comment (fsologureng) [reply templates/repo/issue/view_content.tmpl]: > probably should be `comment` AFAICS

comment role should be for the whole comment container (line 29)

**Review by delvh — APPROVED** [review]
I agree with @zeripath.
I'm not really able to determine it either, so it's better than nothing.

**Comment by jolheiser** [comment]
I would like to see if anything comes of https://github.com/go-gitea/gitea/pull/22612#discussion_r1089000923, but otherwise LGTM

File: templates/repo/issue/view_content.tmpl
Comment (Menelion) [reply templates/repo/issue/view_content.tmpl]: Exactly, @fsologureng is totally right. That's btw how it's done on Github: the whole `username commented...` thing is wrapped into a heading.  
The `article` is good but not enough for navigation. Navigation by headings is in the blood of most blind web users, and a page that lacks headings is extremely inconvenient to navigate through. If you have an `article` with the comment, you have to: go to the article, then arrow up, i.e., back to see the author, then arrow down again past the author to read the comment text now. This is super inconvenient as it requires moving back and forth in the same text. Please note that blind users A) don't use mouse, only keyboard, and B) we navigate in a _linear_ way, we always arrow up and down (we also use the *Tab* key, but it's out of this scope).

File: templates/repo/issue/view_content.tmpl
Comment (Menelion) [reply templates/repo/issue/view_content.tmpl]: @fsologureng Gitea is already overusing the `menu` role, it's full of unlabeled menus. Actually what I would do is to combine the reactions with the comment itself, so it would be in the same `article`, but I'm not confident enough that if I wrap the whole thing in one more `div`, it won't upset the visuals.  
First I worried that I didn't know about the `comment` role, but then I saw why. It's actually in a _draft_ state for ARIA 1.3, meaning that probably no screen reader supports it yet. And yes, even when it gets out of draft, it would be nice to mark the whole comment up with it.  
I agree that `complementary` is not the best fit for reactions. What about `note` then?

Comment (delvh) [reply templates/repo/issue/view_content.tmpl]: > A `note` role suggests a section whose content is parenthetic or ancillary to the main content.

Yep, sounds fine.

Comment (fsologureng) [reply templates/repo/issue/view_content.tmpl]: > @fsologureng Gitea is already overusing the `menu` role, it's full of unlabeled menus. Actually what I would do is to combine the reactions with the comment itself, so it would be in the same `article`, but I'm not confident enough that if I wrap the whole thing in one more `div`, it won't upset the visuals. First I worried that I didn't know about the `comment` role, but then I saw why. It's actually in a _draft_ state for ARIA 1.3, meaning that probably no screen reader supports it yet. And yes, even when it gets out of draft, it would be nice to mark the whole comment up with it. I agree that `complementary` is not the best fit for reactions. What about `note` then?

Yes @Menelion, that bad design of menus should change. I have a PR repairing the label issue of reactions, but is not ready yet. I think both changes could be complementary.

Understood your apprehension about `comment`, I didn't aware of that.

`note` is for content and for complement main resource content. Reactions, instead, are essentially a set of actionable things. What about [`group`](https://www.w3.org/TR/wai-aria-1.1/#group), which is more generic but allows a differentiation with the rest of the comment body. I think it should works.

File: templates/repo/issue/view_content.tmpl
Comment (fsologureng) [reply templates/repo/issue/view_content.tmpl]: How important it is that designers understand this.

**Comment by zeripath** [comment]
@Menelion have you had more thoughts about using `note` instead of `complementary` here? Or do you think this is ready for merge?

**Comment by wolfogre** [comment]
@Menelion Hello, I just wanted to start by saying how much I admire you.

And I noticed that you haven't replied in a long time, so the maintainers haven't merged this PR yet. Your feedback is important to us. I see there are different ideas about the word "complementary", "note" or "group". Considering that there are already two approvals, maybe it shouldn't be blocked for such a long time.

So I'm going to resolve the conflicts and merge this PR after a day if no one else objects. If anyone has a better idea to improve the code, another PR is welcome.

**Comment by Menelion** [comment]
@wolfogre Sorry for such a great delay. I resolved the conflict also (hopefully it was correct).  
I have a question: will it be applied to pull request comments also? Sorry, not so familiar with the project structure yet.

**Comment by delvh** [comment]
> will it be applied to pull request comments also?

What exactly do you mean with `it`?

**Comment by wolfogre** [comment]
> I have a question: will it be applied to pull request comments also?

I believe you are confused about whether it works for both issue pages and PR pages when you only change the files under `repo/issue/`. The answer is yes, as they share the templates you improved.

Thanks! And I'm looking forward to receiving more help from you for Gitea.

**Pull request merged**
